### PR TITLE
assign only the topics in subscription list

### PR DESCRIPTION
**Pull request by zmstone**

### CODE REVIEW

File: src/brod_group_subscriber.erl [hunk 1]
```diff
@@ -35,7 +35,9 @@
 %%%=============================================================================
 
 -module(brod_group_subscriber).
+
 -behaviour(gen_server).
+-behaviour(brod_group_controller).
```
Comment (id) [inline src/brod_group_subscriber.erl]: Behaviour name must be brod_group_subscriber.

I suggest to rename this module to brod_group_subscriber_impl and move behaviour callbacks definition to a new brod_group_subscriber.erl.

Comment (zmstone) [inline src/brod_group_subscriber.erl]: Why 'must'?
a group subscriber is a message subscriber, a group controller is the membership coordinator.
A group subscriber should implement certain callbacks for group controller to interact with.

This doesn't necessary to be a behaviour, having them explicitly declared is easier for compiler/dialyzer checks.

If this only about the name being misleading, how about rename controller to coordinator ?

File: src/brod_group_coordinator.erl [hunk 2]
```diff
@@ -272,12 +284,12 @@ handle_info(?LO_CMD_STABILIZE(N, Reason), State) ->
...
   {ok, NewState} = stabilize(State, 0, {sockent_down, Reason}),
   {noreply, NewState};
-handle_info({'EXIT', Pid, Reason}, #state{subscriber = Pid} = State) ->
+handle_info({'EXIT', Pid, Reason}, #state{member_pid = Pid} = State) ->
   case Reason of
     shutdown      -> {stop, shutdown, State};
     {shutdown, _} -> {stop, shutdown, State};
     normal        -> {stop, normal, State};
-    _             -> {stop, subscriber_down, State}
+    _             -> {stop, member_down, State}
```
Comment (id) [inline src/brod_group_coordinator.erl:313]: why not include Reason here?

File: src/brod_group_coordinator.erl [hunk 3]
```diff
@@ -859,20 +893,21 @@ start_heartbeat_timer(HbRateSec) ->
   ok.
 
 %% @private Start a timer to send a loopback command to self() to trigger
-%% a offset commit request to group coordinator.
+%% a offset commit request to group coordinator borker.
```
Comment (id) [inline src/brod_group_coordinator.erl]: broker :)

File: src/brod_group_coordinator.erl (hunk 2 above)
Comment (zmstone) [inline src/brod_group_coordinator.erl:313]: the process can have its own termination log
also SASL should log crash log.

File: src/brod.erl [hunk 4]
```diff
@@ -324,18 +326,28 @@ subscribe(Client, SubscriberPid, Topic, Partition, Options) ->
 subscribe(ConsumerPid, SubscriberPid, Options) ->
   brod_consumer:subscribe(ConsumerPid, SubscriberPid, Options).
 
-%% @doc Ubsubscribe the current subscriber.
+%% @doc Ubsubscribe the current subscriber. Assuming the subscriber is self().
```
Comment (id) [inline src/brod.erl]: Unsubscribe

File: src/brod.erl [hunk 5]
```diff
@@ -324,18 +326,28 @@ subscribe(Client, SubscriberPid, Topic, Partition, Options) ->
...
 -spec unsubscribe(client(), topic(), partition()) -> ok | {error, any()}.
 unsubscribe(Client, Topic, Partition) ->
+  unsubscribe(Client, Topic, Partition, self()).
+
+%% @doc Ubsubscribe the current subscriber.
```
Comment (id) [inline src/brod.erl]: Unsubscribe

File: src/brod_consumer.erl [hunk 6]
```diff
@@ -145,8 +145,9 @@ subscribe(Pid, SubscriberPid, ConsumerOptions) ->
   safe_gen_call(Pid, {subscribe, SubscriberPid, ConsumerOptions}, infinity).
 
 %% @doc Unsubscribe the current subscriber.
--spec unsubscribe(pid()) -> ok.
-unsubscribe(Pid) -> safe_gen_call(Pid, unsubscribe, infinity).
+-spec unsubscribe(pid(), pid()) -> ok.
```
Comment (id) [inline src/brod_consumer.erl]: ok | {error, any()}

File: src/brod_group_coordinator.erl [hunk 7]
```diff
@@ -544,38 +574,50 @@ sync_group(#state{ groupId      = GroupId
...
-  #partition_assignment{ partition    = Partition
-                       , begin_offset = BeginOffset
-                       , metadata     = Metadata
-                       } = PA,
-  [ io_lib:format("~n    partition=~p begin_offset=~p metadata='~s'",
-                  [Partition, BeginOffset, Metadata])
-  , format_partition_assignments(Rest)].
+-spec merge_acked_offsets(Offsets, Offsets) -> Offsets when
+        Offsets :: [{{topic(), partition()}, offset()}].
+merge_acked_offsets([], ExtraOffsets) -> ExtraOffsets;
```
Comment (id) [inline src/brod_group_coordinator.erl]: wouldn't lists:keymerge/3 do the same?

File: src/brod_group_member.erl [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+%%%    @see assignments_received/4. callback.
+%%% 3. Receive messages from subscribed partitions (delivered by the partition
+%%%    workers (the pollers) implemented in brod_consumer);
+%%% 4. Unsubscribe from all previously subscribed partitions when
+%%%    @see assignments_revoked/1. is called.
+%%%
+%%% For group members who commit offsets to kafka, they should:
+%%% 1. Call @see brod_group_coordinator:ack/4. to acknowledge sucessfull
+%%%    consumption of the messages. Group coordinator will commit the
+%%%    acknwledged offsets every configured interval.
```
Comment (id) [inline src/brod_group_member.erl]: acknowledged

**Comment by zmstone** [comment]
@id merge ?

**Comment by id** [comment]
Sorry, not getting notifications about new commits for some reason.
Let me go through this one more time, I'll reply today.

**Comment by id** [comment]
Amazing work 👍

**Pull request merged**